wicket/src/widgets/status_bar.rs

```diff
@@ -0,0 +1,190 @@
+            // Wicketd is polled every 500ms by wicket. Setting a 1 second
+            // liveness threshold means that under normal operation it should
+            // never flip to delayed.
+            wicketd_liveness: LivenessState::new(Duration::from_secs(1)),
+            // MGS is polled every 10 seconds by wicketd. Set a 11 second
+            // threshold to account for wicket -> wicketd and wicketd -> MGS
+            // delay.
+            mgs_liveness: LivenessState::new(Duration::from_secs(11)),
+            last_redraw_at: None,
+        }
+    }
+
+    pub fn reset_wicketd(&mut self, elapsed: Duration) {
+        self.wicketd_liveness.reset(elapsed);
+        // Force a redraw.
+        self.last_redraw_at = None;
+    }
+
+    pub fn reset_mgs(&mut self, elapsed: Duration) {
+        self.mgs_liveness.reset(elapsed);
+        // Force a redraw.
+        self.last_redraw_at = None;
+    }
+
+    /// Returns true if a redraw needs to happen, resetting the internal timer.
+    pub fn should_redraw(&mut self) -> bool {
+        if let Some(instant) = &mut self.last_redraw_at {
+            let elapsed = instant.elapsed();
+            if elapsed >= Duration::from_secs(1) {
+                *instant = Instant::now();
+                true
+            } else {
+                false
+            }
+        } else {
+            // Initialize the last-redraw timer.
+            self.last_redraw_at = Some(Instant::now());
+            true
+        }
+    }
+
+    /// Returns a widget which can be used to draw the status bar.
+    pub fn to_widget(&self) -> StatusBarWidget<'_> {
+        StatusBarWidget { bar: self }
+    }
+
+    /// Draws the status bar on the bottom of the screen.
+    pub fn draw(&self, f: &mut Frame<'_>) {
+        let widget = self.to_widget();
+        let mut rect = f.size();
+
+        // Draw in the bottom row.
+        rect.y = rect.height - 1;
+        rect.height = 1;
+
+        f.render_widget(widget, rect);
+    }
+}
+
+/// The status bar as a widget.
+pub struct StatusBarWidget<'a> {
+    bar: &'a StatusBar,
+}
+
+impl<'a> Widget for StatusBarWidget<'a> {
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        // "delayed (30s)" is 13 characters
+        const LIVENESS_CELL_WIDTH: u16 = 13;
+
+        let (wicketd_spans, wicketd_spans_width) = center_pad(
+            self.bar.wicketd_liveness.compute().to_spans(),
+            LIVENESS_CELL_WIDTH,
+        );
+        let (mgs_spans, mgs_spans_width) = center_pad(
+            self.bar.mgs_liveness.compute().to_spans(),
+            LIVENESS_CELL_WIDTH,
+        );
+
+        // Render the status bar as a table, using a single row.
+        let row = Row::new(vec![
+            "wicketd".into(),
+            wicketd_spans,
+            // A bit of spacing between wicketd and MGS.
+            " ".into(),
+            "MGS".into(),
+            mgs_spans,
+        ]);
+
+        let table_style = Style::default().fg(OX_GRAY).bg(OX_GREEN_DARKEST);
```

Comment:
While I'm not looking to change it right now, It's nice to have the style settable outside the widget rather than hardcoded, so we can experiment in the caller.  This is typically done with a default / builder interface.

Reply:
Ended up doing a bit of a compromise: styles are defined in a stylesheet, but the stylesheet is hardcoded in the status bar widget currently. (It'll be easy to extract that out if we do per-screen status bar stylesheets, though I think having the same style for the status bar throughout provides uniformity which is nice.)